### PR TITLE
executor, expression: support chunk for HashJoinExec

### PR DESCRIPTION
**Pull request by XuHuaiyu**
to #5261 
I'm testing the memory usage of this implementation.

### CODE REVIEW

File: executor/join.go [hunk 1]
```diff
@@ -271,14 +502,28 @@ func (e *HashJoinExec) prepare(goCtx goctx.Context) error {
...
+	go e.waitJoinWorkersAndCloseResultChan(false)
 	return nil
 }
 
-func (e *HashJoinExec) waitJoinWorkersAndCloseResultChan() {
+// prepare runs the first time when 'Next' is called,
+// it first starts one goroutine to reads all data from the small table to build a hash table,
+// then starts one worker goroutine to fetch rows/chunk from the big table,
+// and, then starts multiple join worker goroutines.
+func (e *HashJoinExec) prepare(goCtx goctx.Context, forChunk bool) error {
```
Comment (zz-jason) [inline executor/join.go]: it's logic is so simple,  I think we can remove it.

File: executor/join.go [hunk 2]
```diff
@@ -209,9 +253,196 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+//	|  joinChkResource[0] |  joinChkResource[1] ...   joinChkResource[n]  |
+//	|        ^            |   ^                            ^              |
+//	|        |            |   |                            |              |
+//	+--------o------------+---o--->joinResultCh<-----------o--------------+
+//	         |                |       +                    |
+//	         |                |       |                    |
+//	         |                |       v                    |
+//	         +----------------+--+NextChunk+---------------+
+
+func (e *HashJoinExec) prepare4Chunk(goCtx goctx.Context) (err error) {
```
Comment (zz-jason) [inline executor/join.go]: the code logical is so simple, I think we can inline it to `NextChunk`

File: executor/join.go [hunk 3]
```diff
@@ -209,9 +253,196 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+	if err = e.build(goCtx); err != nil {
+		return errors.Trace(err)
+	}
+	e.probe(goCtx)
+	return nil
+}
+
+func (e *HashJoinExec) fetchInnerResults(goCtx goctx.Context) (err error) {
```
Comment (zz-jason) [inline executor/join.go]: 1. how about `s/fetchInnerResults/fetchSelectedInners/` ?
2. add a comment, like: `fetchInnerResults fetchs the selected inner rows and copies them to "e.innerResult"`

File: executor/join.go [hunk 4]
```diff
@@ -209,9 +253,196 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+		matched := false
+		selected, matched, err = expression.VectorizedFilter(e.ctx, e.innerFilter, innerExecChk, nil, selected)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if !matched {
+			continue
+		}
+		for idx, ok := range selected {
+			if !ok {
```
Comment (zz-jason) [inline executor/join.go]:
```go
if ok {
    innerResult.AppendRow(innerExecChk.GetRow(idx))
}
```

File: executor/join.go [hunk 5]
```diff
@@ -843,3 +1243,28 @@ func (e *ApplyJoinExec) Next(goCtx goctx.Context) (Row, error) {
 		e.cursor = 0
 	}
 }
+
+func buildMapForList(keyColIdx []int, keyColTypes []*types.FieldType, list *chunk.List, resultMap *mvmap.MVMap, encodeFunc joinKeyEncodeFunc) error {
```
Comment (zz-jason) [inline executor/join.go]: inline it to `buildHashTable` will make code simplified.

File: executor/join.go [hunk 6]
```diff
@@ -164,6 +192,22 @@ func getJoinKey(cols []*expression.Column, row Row, vals []types.Datum, bytes []
 	return false, bytes, errors.Trace(err)
 }
 
+type joinKeyEncodeFunc func(b []byte, k types.Datum) (bool, []byte, error)
+
+func getJoinKeyFromChkRow(keyColIdx []int, keyColTypes []*types.FieldType, row chunk.Row, keyBuf []byte, encodeFunc joinKeyEncodeFunc) (bool, []byte, error) {
+	keyBuf = keyBuf[:0]
+	for i, colIdx := range keyColIdx {
+		d := row.GetDatum(colIdx, keyColTypes[i])
```
Comment (zz-jason) [inline executor/join.go]: I think we'd better not to call `row.GetDatum(colIdx, keyColTypes[i])`, it's inefficient and will be deprecated in the near future.

File: executor/join.go [hunk 7]
```diff
@@ -209,9 +253,196 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+				continue
+			}
+			innerResult.AppendRow(innerExecChk.GetRow(idx))
+		}
+	}
+	e.innerResult = innerResult
+	return nil
+}
+
+func (e *HashJoinExec) getJoinKeyEncodeFunc() joinKeyEncodeFunc {
```
Comment (zz-jason) [inline executor/join.go]: no need to do it this way

File: executor/join.go [hunk 8]
```diff
@@ -209,9 +253,196 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+	for i := 0; i < e.concurrency; i++ {
+		e.joinChkResource[i] = make(chan *chunk.Chunk, e.concurrency)
+		e.joinChkResource[i] <- e.newChunk()
+	}
+
+	e.joinResultCh = make(chan *execWorkerResult, e.concurrency)
+	go e.waitJoinWorkersAndCloseResultChan(true)
+}
+
+func (e *HashJoinExec) probe(goCtx goctx.Context) {
```
Comment (zz-jason) [inline executor/join.go]: how about `s/probe/fetchOuterAndProbeHashTable` ?

File: executor/join.go [hunk 9]
```diff
@@ -209,9 +255,147 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+			return true, nil, nil
+		}
+		var err error
+		b, err = codec.HashValues(b, k)
+		return false, b, errors.Trace(err)
+	}
+}
+
+func (e *HashJoinExec) build(goCtx goctx.Context) (err error) {
+	e.hashTable = mvmap.NewMVMap()
```
Comment (winoros) [inline executor/join.go]: this is useless?

File: executor/join.go [hunk 10]
```diff
@@ -208,9 +261,128 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+		outerResult := &execWorkerResult{
+			src: e.outerChkResource,
+		}
+
+		select {
+		case _, ok := <-e.closeCh:
+			if !ok {
+				return
+			}
+		case outerResult.chk = <-e.outerChkResource:
```
Comment (zz-jason) [inline executor/join.go]: We can make `e.outerChkResource` be a `chan` of `execWorkerResult`, once we received a `execWorkerResult`, we fetch the outer chunk and put that chunk into `execWorkerResult.src`.

This is more precise to decide which channel we should put the outer chunk into.

File: executor/join.go [hunk 11]
```diff
@@ -208,9 +263,118 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+	for i := 0; i < e.concurrency; i++ {
+		e.outerChkResourceCh <- &execWorkerResult{
+			chk: e.outerExec.newChunk(),
+			src: e.outerResultChs[i],
+		}
+	}
+
+	e.joinChkResourceCh = make([]chan *chunk.Chunk, e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		e.joinChkResourceCh[i] = make(chan *chunk.Chunk, e.concurrency)
```
Comment (zz-jason) [inline executor/join.go]:
```go
e.joinChkResourceCh[i] = make(chan *chunk.Chunk, 1)
```

**Comment by XuHuaiyu** [comment]
PTAL @coocood

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason @coocood

File: executor/executor.go [hunk 12]
```diff
@@ -749,7 +749,10 @@ func (e *SelectionExec) NextChunk(goCtx goctx.Context, chk *chunk.Chunk) error {
 			if chk.NumRows() == e.maxChunkSize {
 				return nil
 			}
-			chk.AppendRow(e.inputRow)
+			chk.AppendPartialRow(0, e.inputRow)
```
Comment (zz-jason) [inline executor/executor.go]: now we can
```go
chk.AppendRow(e.inputRow)
```

File: executor/join.go [hunk 13]
```diff
@@ -406,18 +626,122 @@ func (e *HashJoinExec) joinOuterRow(workerID int, outerRow Row, resultBuffer *ex
...
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (e *HashJoinExec) buildNewJoinResult(workerID int, joinResult *execWorkerResult) (bool, *execWorkerResult) {
+	joinResult = &execWorkerResult{
+		cycleChan: e.joinChkResourceCh[workerID],
+	}
+	ok := true
```
Comment (zz-jason) [inline executor/join.go]: we can:
```go
select {
    case <-e.closeCh:
        ok = false
    case joinResult.chk, ok = <-e.joinChkResourceCh[workerID]:
}
return ok, joinResult
```

File: executor/join.go [hunk 14]
```diff
@@ -371,6 +538,59 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
...
+	select {
+	case <-e.closeCh:
+		return
+	case joinResult.chk, ok = <-e.joinChkResourceCh[workerID]:
+		if !ok {
+			return
+		}
+	}
+	// Read and filter outerResult, and join the outerResult with the inner rows.
+	joinResultChkBuffer := e.newChunk()
```
Comment (zz-jason) [inline executor/join.go]: s/joinResultChkBuffer/joinResultBuffer/g

File: executor/join.go [hunk 15]
```diff
@@ -371,6 +538,59 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
...
+	var (
+		outerResult *chunk.Chunk
+		joinResult  *execWorkerResult
+		selected    = make([]bool, 0, chunk.InitialCapacity)
+		ok          bool
+	)
+	joinResult = &execWorkerResult{
+		cycleChan: e.joinChkResourceCh[workerID],
+	}
+	select {
```
Comment (zz-jason) [inline executor/join.go]:
```go
ok, joinResult := e.buildNewJoinResult(workerID)
if !ok {
    return
}
```

File: executor/join.go [hunk 16]
```diff
@@ -371,6 +538,59 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
 	e.workerWaitGroup.Done()
 }
 
+func (e *HashJoinExec) runJoinWorker4Chunk(workerID int) {
+	defer func() {
```
Comment (zz-jason) [inline executor/join.go]:
```go
defer e.workerWaitGroup.Done()
```

File: executor/join.go [hunk 17]
```diff
@@ -406,18 +626,122 @@ func (e *HashJoinExec) joinOuterRow(workerID int, outerRow Row, resultBuffer *ex
...
+	}
+
+	err = e.resultGenerators[workerID].emitToChunk(outerRow, chunk.NewSliceIterator(innerRows), chk)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (e *HashJoinExec) buildNewJoinResult(workerID int, joinResult *execWorkerResult) (bool, *execWorkerResult) {
```
Comment (zz-jason) [inline executor/join.go]:
```go
buildNewJoinResult(workerID int) (bool, *execWorkerResult) {
```

File: executor/join.go [hunk 18]
```diff
@@ -406,18 +626,122 @@ func (e *HashJoinExec) joinOuterRow(workerID int, outerRow Row, resultBuffer *ex
...
+			if err != nil {
+				joinResult.err = errors.Trace(err)
+				return false, joinResult
+			}
+		}
+		// Splitting the joinResultChkBuffer into chunks that reach e.maxChunkSize.
+		if joinResultChkBuffer.NumRows() == 0 {
+			continue
+		}
+		startRowOffset := e.maxChunkSize - joinResult.chk.NumRows()
```
Comment (zz-jason) [inline executor/join.go]:
```go
numRemained := e.maxChunkSize - joinResult.chk.NumRows()
numAppended := mathutil.Min(joinResultChkBuffer.NumRows(), numRemained)
```

File: executor/join.go [hunk 19]
```diff
@@ -406,18 +626,122 @@ func (e *HashJoinExec) joinOuterRow(workerID int, outerRow Row, resultBuffer *ex
...
+		joinResult.chk.Append(joinResultChkBuffer, 0, mathutil.Min(joinResultChkBuffer.NumRows(), startRowOffset))
+		if joinResult.chk.NumRows() < e.maxChunkSize {
+			continue
+		}
+		e.joinResultCh <- joinResult
+		ok, joinResult = e.buildNewJoinResult(workerID, joinResult)
+		if !ok {
+			return false, joinResult
+		}
+		if cnt := (joinResultChkBuffer.NumRows() - startRowOffset) / e.maxChunkSize; cnt >= 1 {
```
Comment (zz-jason) [inline executor/join.go]:
```go
for numAppended < joinResultChkBuffer.NumRows() {
    numBatchSize := mathutil.Min(e.maxChunkSize, joinResultChkBuffer.NumRows()-numAppended)
    ok, joinResult = e.buildNewJoinResult(workerID, joinResult)
    if !ok {
        return false, joinResult
    }
    joinResult.chk.Append(joinResultChkBuffer, numAppended, numAppended+numBatchSize)
    numAppended += numBatchSize
}
```

File: executor/join.go [hunk 20]
```diff
@@ -406,18 +626,122 @@ func (e *HashJoinExec) joinOuterRow(workerID int, outerRow Row, resultBuffer *ex
...
+}
+
+func (e *HashJoinExec) join2Chunk(workerID int, outerChk *chunk.Chunk, joinResultChkBuffer *chunk.Chunk, joinResult *execWorkerResult, selected []bool) (ok bool, _ *execWorkerResult) {
+	var err error
+	selected, err = expression.VectorizedFilter(e.ctx, e.outerFilter, outerChk, selected)
+	if err != nil {
+		joinResult.err = errors.Trace(err)
+		return false, joinResult
+	}
+	for i, matched := range selected {
```
Comment (zz-jason) [inline executor/join.go]: no need to bring in `matched`, it increases the difficulty to understand the code behavior. 
```go
for i := range selected
```

File: executor/join.go [hunk 21]
```diff
@@ -209,9 +264,118 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+			return errors.Trace(err)
+		}
+		if innerExecChk.NumRows() == 0 {
+			break
+		}
+		selected, err = expression.VectorizedFilter(e.ctx, e.innerFilter, innerExecChk, selected)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for idx, ok := range selected {
```
Comment (zz-jason) [inline executor/join.go]: ditto, `for i := range selected {` is enough

File: executor/join.go [hunk 22]
```diff
@@ -209,9 +264,118 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+				innerResult.AppendRow(innerExecChk.GetRow(idx))
+			}
+		}
+	}
+	e.innerResult = innerResult
+	return nil
+}
+
+func (e *HashJoinExec) initializeForProbe() {
+	e.outerResultChs = make([]chan *chunk.Chunk, e.concurrency)
```
Comment (zz-jason) [inline executor/join.go]: for all the channels created in this function, maybe it's better for understanding to add comments for each of them to explain who will read them and who will write them ?

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason

File: executor/join.go [hunk 23]
```diff
@@ -209,9 +282,124 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+			}
+		}
+	}
+	e.innerResult = innerResult
+	return nil
+}
+
+func (e *HashJoinExec) initializeForProbe() {
+	// e.outerResultChs is for transmitting the chunks which store the data of outerExec,
+	// it'll be written by outer worker goroutine, and read be join workers.
```
Comment (zz-jason) [inline executor/join.go]: s/be/by/

File: executor/join.go [hunk 24]
```diff
@@ -371,6 +562,51 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
 	e.workerWaitGroup.Done()
 }
 
+func (e *HashJoinExec) runJoinWorker4Chunk(workerID int) {
+	defer e.workerWaitGroup.Done()
+	var (
```
Comment (zz-jason) [inline executor/join.go]: line 569 and 571 can be deleted

File: executor/join.go [hunk 25]
```diff
@@ -209,9 +282,124 @@ func (e *HashJoinExec) fetchOuterRows(goCtx goctx.Context) {
...
+		e.workerWaitGroup.Done()
+	}()
+	for {
+		if e.finished.Load().(bool) {
+			return
+		}
+		var outerResource *outerChkResource
+		ok := true
+		select {
+		case <-e.closeCh:
```
Comment (coocood) [inline executor/join.go]: Can this case be removed?
When the executor is closed the outerChkResouceCh will be closed.

File: executor/join.go [hunk 26]
```diff
@@ -371,6 +562,49 @@ func (e *HashJoinExec) runJoinWorker(workerID int) {
...
+	joinResultBuffer := e.newChunk()
+	emptyOuterResult := &outerChkResource{
+		dest: e.outerResultChs[workerID],
+	}
+	for ok := true; ok; {
+		if e.finished.Load().(bool) {
+			break
+		}
+		select {
+		case <-e.closeCh:
```
Comment (coocood) [inline executor/join.go]: ditto

**Comment by XuHuaiyu** [comment]
PTAL @coocood

**Pull request merged**